File: packages/compiler/src/assets.ts
import ts from 'typescript';
import type * as nativeTs from './native-typescript.js';
import type {
	ExactAssetDependencyIR,
	ExactAssetImportMode,
	ExactAssetRule,
	ExactAssetTarget
} from './types.js';

/** Defines the exact import placement type contract. */
export type ExactImportPlacement = 'client' | 'server';

/** Describes the planned exact module import operation. */
export type ExactModuleImportPlan = {
	readonly assets: readonly ExactAssetDependencyIR[];
	readonly diagnostics: readonly string[];
	readonly placementBySpecifier: ReadonlyMap<string, ExactImportPlacement>;
	readonly clientAssetSideEffectStarts: ReadonlySet<number>;
};

const defaultAssetRules: readonly ExactAssetRule[] = [
	{
		extensions: ['.css', '.less', '.scss'],
		kind: 'style',
		deliveryTarget: 'client'
	}
];

/** Computes a module imports without changing caller-owned input. */
export function analyzeModuleImports(
	source: string,
	filename: string,
	rules: readonly ExactAssetRule[] = []
): ExactModuleImportPlan {
	const sourceFile = ts.createSourceFile(
		filename,
		source,
		ts.ScriptTarget.ES2022,
		true,
		scriptKind(filename)
	);
	const assets: ExactAssetDependencyIR[] = [];
	const diagnostics: string[] = [];
	const placementBySpecifier = new Map<string, ExactImportPlacement>();
	const clientAssetSideEffectStarts = new Set<number>();

	for (const statement of sourceFile.statements) {
		if (!ts.isImportDeclaration(statement) || !ts.isStringLiteral(statement.moduleSpecifier))
			continue;
		const specifier = statement.moduleSpecifier.text;
		const placement = exactImportPlacement(statement, sourceFile, diagnostics);
		if (placement) {
			const prior = placementBySpecifier.get(specifier);
			if (prior && prior !== placement)
				diagnostics.push(`error: import '${specifier}' has conflicting exact placement attributes`);
			else placementBySpecifier.set(specifier, placement);
		}

		const rule = matchingAssetRule(specifier, [...rules, ...defaultAssetRules]);
		if (!rule) continue;
		const sideEffect = !statement.importClause;
		const importMode = rule.importMode ?? (sideEffect ? 'side-effect' : 'module');
		const evaluationTarget =
			placement ?? rule.evaluationTarget ?? defaultEvaluationTarget(importMode, sideEffect);
		const deliveryTarget = rule.deliveryTarget ?? defaultDeliveryTarget(importMode);
		assets.push({ specifier, kind: rule.kind, importMode, evaluationTarget, deliveryTarget });
		if (sideEffect && evaluationTarget === 'client')
			clientAssetSideEffectStarts.add(statement.getStart(sourceFile));
		if (!placement && evaluationTarget !== 'both')
			placementBySpecifier.set(specifier, evaluationTarget);
	}

	return Object.freeze({
		assets: Object.freeze(dedupeAssets(assets)),
		diagnostics: Object.freeze([...new Set(diagnostics)]),
		placementBySpecifier,
		clientAssetSideEffectStarts
	});
}

/** Performs the strip exact import attribute domain operation. */
export function stripExactImportAttribute(
	node: nativeTs.ImportDeclaration,
	factory: nativeTs.NodeFactory
): nativeTs.ImportDeclaration {
	const attributes = node.attributes;
	if (!attributes) return node;
	const retained = attributes.attributes.filter(
		(element) => importAttributeName(element.name) !== 'exact'
	);
	if (retained.length === attributes.attributes.length) return node;
	const next = retained.length
		? factory.updateImportAttributes(
				attributes,
				factory.createNodeArray(retained),
				attributes.multiLine
			)
		: undefined;
	return factory.updateImportDeclaration(
		node,
		node.modifiers,
		node.importClause,
		node.moduleSpecifier,
		next
	);
}

function exactImportPlacement(
	node: ts.ImportDeclaration,
	sourceFile: ts.SourceFile,
	diagnostics: string[]
): ExactImportPlacement | undefined {
	const matches =
		node.attributes?.elements.filter((element) => importAttributeName(element.name) === 'exact') ??
		[];
	if (!matches.length) return undefined;
	const location = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
	const prefix = `${sourceFile.fileName}:${location.line + 1}:${location.character + 1}`;
	if (matches.length > 1)
		diagnostics.push(`error: ${prefix} import has duplicate exact attributes`);
	if (node.importClause?.isTypeOnly)
		diagnostics.push(`error: ${prefix} type-only import cannot declare exact placement`);
	const value = matches[0]?.value;
	if (
		!value ||
		!ts.isStringLiteral(value) ||
		(value.text !== 'client' && value.text !== 'server')
	) {
		diagnostics.push(`error: ${prefix} exact import attribute must be \"client\" or \"server\"`);
		return undefined;
	}
	return value.text;
}

function importAttributeName(name: Readonly<{ text: string }>): string {
	return name.text;
}

function matchingAssetRule(
	specifier: string,
	rules: readonly ExactAssetRule[]
): ExactAssetRule | undefined {
	const [pathname, query = ''] = specifier.split('?', 2);
	const clean = pathname!.split('#', 1)[0]!.toLowerCase();
	const queryTokens = new Set(
		query
			.split('&')
			.map((token) => token.split('=', 1)[0]!.toLowerCase())
			.filter(Boolean)
	);
	return rules.find((rule) => {
		const extensionMatch =
			!rule.extensions?.length ||
			rule.extensions.some((extension) => clean.endsWith(normalizeExtension(extension)));
		const queryMatch =
			!rule.queries?.length ||
			rule.queries.some((token) => queryTokens.has(token.replace(/^\?/, '').toLowerCase()));
		return extensionMatch && queryMatch;
	});
}

function normalizeExtension(extension: string): string {
	const normalized = extension.toLowerCase();
	return normalized.startsWith('.') ? normalized : `.${normalized}`;
}

function defaultEvaluationTarget(
	mode: ExactAssetImportMode,
	sideEffect: boolean
): Exclude<ExactAssetTarget, 'embedded'> {
	if (mode === 'worker' || sideEffect) return 'client';
	return 'both';
}

function defaultDeliveryTarget(mode: ExactAssetImportMode): ExactAssetTarget {
	return mode === 'raw' || mode === 'inline' ? 'embedded' : 'client';
}

function dedupeAssets(values: readonly ExactAssetDependencyIR[]): ExactAssetDependencyIR[] {
	const seen = new Set<string>();
	return values.filter((value) => {
		const key = JSON.stringify(value);
		if (seen.has(key)) return false;
		seen.add(key);
		return true;
	});
}

function scriptKind(filename: string): ts.ScriptKind {
	if (/\.tsx$/i.test(filename)) return ts.ScriptKind.TSX;
	if (/\.jsx$/i.test(filename)) return ts.ScriptKind.JSX;
	if (/\.js$/i.test(filename)) return ts.ScriptKind.JS;
	return ts.ScriptKind.TS;
}

File: native/typescript-go/overlay/internal/exactcompiler/assets.go
package exactcompiler

import (
	"fmt"
	"strings"

	"github.com/microsoft/typescript-go/internal/ast"
	"github.com/microsoft/typescript-go/internal/checker"
	"github.com/microsoft/typescript-go/internal/printer"
)

type assetAnalysis struct {
	dependencies           []AssetDependency
	placementBySpecifier   map[string]string
	clientSideEffectStarts map[int]struct{}
	omitStatementStarts    map[string]map[int]struct{}
	diagnostics            []Diagnostic
}

var nativeDefaultAssetRules = []AssetRule{{
	Extensions:     []string{".css", ".less", ".scss"},
	Kind:           "style",
	DeliveryTarget: "client",
}}

func analyzeAssets(
	sourceFile *ast.SourceFile,
	typeChecker *checker.Checker,
	request Request,
) assetAnalysis {
	result := assetAnalysis{
		dependencies:           []AssetDependency{},
		placementBySpecifier:   make(map[string]string),
		clientSideEffectStarts: make(map[int]struct{}),
		omitStatementStarts: map[string]map[int]struct{}{
			"client": {},
			"server": {},
		},
		diagnostics: []Diagnostic{},
	}
	bindingPlacements := make(map[ast.SymbolId]string)
	placedBindingNames := make(map[string]struct{})
	rules := append(append([]AssetRule(nil), request.AssetRules...), nativeDefaultAssetRules...)
	seenAssets := make(map[string]struct{})
	for _, statement := range sourceFile.Statements.Nodes {
		if !ast.IsImportDeclaration(statement) {
			continue
		}
		declaration := statement.AsImportDeclaration()
		if !ast.IsStringLiteral(declaration.ModuleSpecifier) {
			continue
		}
		specifier := declaration.ModuleSpecifier.AsStringLiteral().Text
		placement, placementDiagnostics := exactImportPlacement(
			sourceFile,
			statement,
			declaration,
		)
		result.diagnostics = append(result.diagnostics, placementDiagnostics...)
		if placement != "" {
			if prior := result.placementBySpecifier[specifier]; prior != "" && prior != placement {
				result.diagnostics = append(result.diagnostics, Diagnostic{
					Severity: "error",
					Code:     "EXACT5004",
					Message: fmt.Sprintf(
						"error: import %q has conflicting exact placement attributes",
						specifier,
					),
					Start:  statement.Pos(),
					Length: statement.End() - statement.Pos(),
				})
			} else {
				result.placementBySpecifier[specifier] = placement
			}
			collectPlacedImportSymbols(
				declaration,
				placement,
				typeChecker,
				bindingPlacements,
				placedBindingNames,
			)
		}
		rule, matched := matchingNativeAssetRule(specifier, rules)
		if !matched {
			continue
		}
		sideEffect := declaration.ImportClause == nil
		importMode := rule.ImportMode
		if importMode == "" {
			if sideEffect {
				importMode = "side-effect"
			} else {
				importMode = "module"
			}
		}
		evaluationTarget := placement
		if evaluationTarget == "" {
			evaluationTarget = rule.EvaluationTarget
		}
		if evaluationTarget == "" {
			if importMode == "worker" || sideEffect {
				evaluationTarget = "client"
			} else {
				evaluationTarget = "both"
			}
		}
		deliveryTarget := rule.DeliveryTarget
		if deliveryTarget == "" {
			if importMode == "raw" || importMode == "inline" {
				deliveryTarget = "embedded"
			} else {
				deliveryTarget = "client"
			}
		}
		dependency := AssetDependency{
			Specifier:        specifier,
			Kind:             rule.Kind,
			ImportMode:       importMode,
			EvaluationTarget: evaluationTarget,
			DeliveryTarget:   deliveryTarget,
		}
		key := fmt.Sprintf("%+v", dependency)
		if _, exists := seenAssets[key]; !exists {
			seenAssets[key] = struct{}{}
			result.dependencies = append(result.dependencies, dependency)
		}
		if sideEffect && evaluationTarget == "client" {
			result.clientSideEffectStarts[statement.Pos()] = struct{}{}
		}
		if placement == "" && evaluationTarget != "both" {
			result.placementBySpecifier[specifier] = evaluationTarget
		}
	}
	if len(bindingPlacements) == 0 {
		return result
	}
	for _, statement := range sourceFile.Statements.Nodes {
		if ast.IsImportDeclaration(statement) {
			continue
		}
		walkNode(statement, func(node *ast.Node) bool {
			if !ast.IsIdentifier(node) || ast.IsDeclarationName(node) ||
				isStaticPropertyName(node) {
				return true
			}
			if _, relevant := placedBindingNames[node.Text()]; !relevant {
				return true
			}
			symbol := typeChecker.GetSymbolAtLocation(node)
			if symbol == nil {
				return true
			}
			placement := bindingPlacements[ast.GetSymbolId(symbol)]
			if placement == "" {
				return true
			}
			opposite := "client"
			if placement == "client" {
				opposite = "server"
			}
			result.omitStatementStarts[opposite][statement.Pos()] = struct{}{}
			return true
		})
	}
	return result
}

func collectPlacedImportSymbols(
	declaration *ast.ImportDeclaration,
	placement string,
	typeChecker *checker.Checker,
	result map[ast.SymbolId]string,
	relevantNames map[string]struct{},
) {
	if declaration.ImportClause == nil {
		return
	}
	clause := declaration.ImportClause.AsImportClause()
	names := []*ast.Node{}
	if clause.Name() != nil {
		names = append(names, clause.Name())
	}
	if clause.NamedBindings != nil {
		if ast.IsNamespaceImport(clause.NamedBindings) {
			names = append(names, clause.NamedBindings.AsNamespaceImport().Name())
		} else {
			for _, specifier := range clause.NamedBindings.AsNamedImports().Elements.Nodes {
				names = append(names, specifier.Name())
			}
		}
	}
	for _, name := range names {
		relevantNames[name.Text()] = struct{}{}
		if symbol := typeChecker.GetSymbolAtLocation(name); symbol != nil {
			result[ast.GetSymbolId(symbol)] = placement
		}
	}
}

func exactImportPlacement(
	sourceFile *ast.SourceFile,
	statement *ast.Node,
	declaration *ast.ImportDeclaration,
) (string, []Diagnostic) {
	if declaration.Attributes == nil {
		return "", nil
	}
	matches := []*ast.Node{}
	for _, attribute := range declaration.Attributes.AsImportAttributes().Attributes.Nodes {
		if attribute.AsImportAttribute().Name().Text() == "exact" {
			matches = append(matches, attribute)
		}
	}
	if len(matches) == 0 {
		return "", nil
	}
	line, column := sourceLocation(sourceFile, statement.Pos())
	prefix := fmt.Sprintf("%s:%d:%d", sourceFile.FileName(), line, column)
	diagnostics := []Diagnostic{}
	add := func(code string, message string) {
		diagnostics = append(diagnostics, Diagnostic{
			Severity: "error",
			Code:     code,
			Message:  "error: " + prefix + " " + message,
			Start:    statement.Pos(),
			Length:   statement.End() - statement.Pos(),
		})
	}
	if len(matches) > 1 {
		add("EXACT5001", "import has duplicate exact attributes")
	}
	if declaration.ImportClause != nil &&
		declaration.ImportClause.AsImportClause().PhaseModifier == ast.KindTypeKeyword {
		add("EXACT5002", "type-only import cannot declare exact placement")
	}
	value := matches[0].AsImportAttribute().Value
	if value == nil || !ast.IsStringLiteral(value) {
		add("EXACT5003", `exact import attribute must be "client" or "server"`)
		return "", diagnostics
	}
	placement := value.AsStringLiteral().Text
	if placement != "client" && placement != "server" {
		add("EXACT5003", `exact import attribute must be "client" or "server"`)
		return "", diagnostics
	}
	return placement, diagnostics
}

func matchingNativeAssetRule(specifier string, rules []AssetRule) (AssetRule, bool) {
	parts := strings.SplitN(specifier, "?", 2)
	clean := strings.ToLower(strings.SplitN(parts[0], "#", 2)[0])
	queryTokens := make(map[string]struct{})
	if len(parts) > 1 {
		for _, value := range strings.Split(parts[1], "&") {
			token := strings.ToLower(strings.SplitN(value, "=", 2)[0])
			if token != "" {
				queryTokens[token] = struct{}{}
			}
		}
	}
	for _, rule := range rules {
		extensionMatch := len(rule.Extensions) == 0
		for _, extension := range rule.Extensions {
			normalized := strings.ToLower(extension)
			if !strings.HasPrefix(normalized, ".") {
				normalized = "." + normalized
			}
			extensionMatch = extensionMatch || strings.HasSuffix(clean, normalized)
		}
		queryMatch := len(rule.Queries) == 0
		for _, query := range rule.Queries {
			_, exists := queryTokens[strings.TrimPrefix(strings.ToLower(query), "?")]
			queryMatch = queryMatch || exists
		}
		if extensionMatch && queryMatch {
			return rule, true
		}
	}
	return AssetRule{}, false
}

func stripExactImportAttribute(
	declaration *ast.ImportDeclaration,
	factory *printer.NodeFactory,
) *ast.Node {
	if declaration.Attributes == nil {
		return declaration.AsNode()
	}
	attributes := declaration.Attributes.AsImportAttributes()
	retained := make([]*ast.Node, 0, len(attributes.Attributes.Nodes))
	for _, attribute := range attributes.Attributes.Nodes {
		if attribute.AsImportAttribute().Name().Text() != "exact" {
			retained = append(retained, attribute)
		}
	}
	if len(retained) == len(attributes.Attributes.Nodes) {
		return declaration.AsNode()
	}
	var updatedAttributes *ast.Node
	if len(retained) != 0 {
		updatedAttributes = factory.UpdateImportAttributes(
			attributes,
			attributes.Token,
			factory.NewNodeList(retained),
			attributes.MultiLine,
		)
	}
	return factory.UpdateImportDeclaration(
		declaration,
		declaration.Modifiers(),
		declaration.ImportClause,
		declaration.ModuleSpecifier,
		updatedAttributes,
	)
}
